### PR TITLE
[FSDP] Fix `_get_param_to_unflat_param_names()` for shared params

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #75431 [FSDP] Add `ignored_modules` ctor arg
* **#75430 [FSDP] Fix `_get_param_to_unflat_param_names()` for shared params**

This removes an `assert` I had when originally implementing the function to check the recursion setup, which is actually violated for shared parameters. This clarifies the shared parameter case.

Differential Revision: [D35550145](https://our.internmc.facebook.com/intern/diff/D35550145)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2468,6 +2468,9 @@ def _get_param_to_unflat_param_names(
     parameters, these mapped-to lists always contain a single element. The
     unflattened parameter names should match the keys of the model state dict.
 
+    For shared parameters, only the first parameter name is included (following
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2471]: This follows the `nn.Module.parameters()` convention and suffices for the current use cases of this function.

Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2471]: Do we know why this didn't come up in unittests? I recall we have a module that uses shared params in transformer module in `common_fsdp.py`. Can we add a unittest to catch this in the future?

**Review by rohan-varma — APPROVED** [review]
stamp to unblock but pls consider adding test

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2471]: In my unit tests, I did exercise code paths that call this helper method on the `TransformerWithSharedParams` *after* wrapping with `FSDP`. The issue arises when calling this method on a non-wrapped `nn.Module` with shared parameters. I only noticed the `assert` failing for the `TransformerWithSharedParams` when using this helper in the `FSDP` constructor *before* wrapping.

I will see how to best incorporate this case into the unit tests. It was certainly an oversight on my end.

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550145).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35550145).

**Pull request closed without merging**